juju/terraform-provider-juju
# Fix application deploy and refresh internals

## Description

Visual inspection of the code before working to add base support for application resources found a number of issues to fix.
* When refreshing, update application constraints before adding new units, otherwise the new units will not use the new constraints.
* During juju deploy, check the application constraints for architecture, fallback to using the model constraint value if not provided.
* Improve operating system selection for deploying a charm if one is not provided. Including validating against juju supported workload operating systems if a juju 3.x controller.
* Updating an applications's charm revision or channel requires revalidation of the charm with new data against the operating system and architecture deployed with. As well as ensuring they exist.
* When refreshing an application, use the current application charm origin to start. Do not deduce a new one. This could lead to the wrong charm in updates as a charm name can change, the origin holds an ID which is immutable and will guarantee the same charm is used after deploy.
* Do not change a charm series after deployed. That could cause incompatible charm blobs to be used. 
* Do not create juju clients if they will not be used. 
* Close a juju api connection, not each individual client created using the connection. Once a client is closed, the shared connection is also closed, potentially preventing use by other clients.

## Type of change

- Logic changes in resources (the API interaction with Juju has been changed)

## QA steps

Please create a variety of applications plans to ensure no change in behavior, modulo the not changing the series on application resource update. Focus on constraints, charm revision/channel, charm refresh, operating system selection. 

Try deploying jameinel-ubuntu-lite, this is where I noticed that the provider could not determine the correct operating system to deploy with as it failed to deploy.

---
The following plan should produce an error based on the juju bug found
```terraform
terraform {
  required_providers {
    juju = {
      version = ">= 0.9.0"
      source  = "juju/juju"
    }
  }
}

provider "juju" {
}

resource "juju_model" "testmodel" {
  name = "focal-test"
}

resource "juju_application" "ubuntu" {
  model = juju_model.testmodel.name
  charm {
    name = "postgresql"
    series = "focal"
  }
}
```

## Additional notes

JUJU-4711
https://bugs.launchpad.net/juju/+bug/2039179
